Gems/Atom/Tools/MaterialCanvas/Code/Source/Document/MaterialCanvasDocument.cpp

```diff
@@ -352,4 +372,337 @@ namespace MaterialCanvas
         GraphCanvas::SceneRequestBus::Event(m_graphId, &GraphCanvas::SceneRequests::ClearScene);
         GraphCanvas::GraphModelRequestBus::Event(m_graphId, &GraphCanvas::GraphModelRequests::RequestPopPreventUndoStateUpdate);
     }
+
+    bool MaterialCanvasDocument::CompareNodeExecutionOrder(GraphModel::ConstNodePtr nodeA, GraphModel::ConstNodePtr nodeB) const
+    {
+        if (!nodeA || !nodeB)
+        {
+            return false;
+        }
+
+        if (nodeA == nodeB)
+        {
+            return true;
+        }
+
+        for (const auto& slotPair : nodeA->GetSlots())
+        {
+            if (slotPair.second->GetSlotDirection() == GraphModel::SlotDirection::Input)
+            {
+                for (const auto& connection : slotPair.second->GetConnections())
+                {
+                    if (CompareNodeExecutionOrder(connection->GetSourceNode(), nodeB))
+                    {
+                        return true;
+                    }
+                }
+            }
+        }
+
+        return false;
+    }
+
+    AZStd::vector<GraphModel::ConstNodePtr> MaterialCanvasDocument::GetNodesInExecutionOrder() const
+    {
+        AZStd::vector<GraphModel::ConstNodePtr> nodes;
+        nodes.reserve(m_graph->GetNodes().size());
+
+        for (const auto& nodePair : m_graph->GetNodes())
+        {
+            nodes.push_back(nodePair.second);
+        }
+
+        AZStd::sort(
```

Comment:
We may want to replace this with a cheaper topological sort later which is linear in $O(N + E)$, with $N$ the number of nodes and $E$ the number of edges. This algorithm appears to be $O(NE^2\log N)$. [This](https://en.wikipedia.org/wiki/Topological_sorting#Kahn's_algorithm) is my favorite approach to this, implementable in a few lines of code.

Reply:
I will look at the algorithm. It should be very easy to replace the sort with something more efficient. Have to revisit the sort anyway to fix an edge case that can occur if one node branches to multiple output connections and one of those branches is a dead end. it can cause a variable to be used before it's declared.